tests/utilities/test_cli.py

```diff
@@ -614,3 +614,117 @@ def test_cli_config_overwrite(tmpdir):
         LightningCLI(BoringModel, trainer_defaults=trainer_defaults)
     with mock.patch('sys.argv', ['any.py']):
         LightningCLI(BoringModel, save_config_overwrite=True, trainer_defaults=trainer_defaults)
+
+
+def test_lightning_cli_optimizer(tmpdir):
```

Comment:
I couldn't see it in the test or docs, but is it possible to configure a categorical choice, something like to easily switch between 

```
--optimizer=sgd

--optimizer=adam
```

for example, or do we always need to give the full dictionary?

Reply:
Yeah, the cateogical choice is what I was getting at in my comment below as well---where there is a default choice and default parameters given.

I think that what @awaelchli  is pointing at is the key:  not sure we have a sense what the current options are for construction and for setting default parameters and teh default optimizer/scheduler are.  If they are in the docs, I missed them.

Reply:
It is not always necessary to give a full dictionary. Doing `--optimizer=torch.optim.Adam` would use `Adam` and whatever defaults it has. The same would work for other classes in `torch.optim` or any other place as long as they inherit from `Optimizer`. However, note that for `SGD` the learning rate by design is a required parameter and has no default. So in this case only giving the class path would not work. Someone could subclass `SGD` to give `lr` a default and make it work. But this is not recommended because for `SGD` there is no good default. The range of its optimal value depends on the task.

Reply:
> However, note that for SGD the learning rate by design is a required parameter and has no default. So in this case only giving the class path would not work. 

Of course.  But that doesn't meant the user should be setting it each time.  YOu find a good one and then you bake it into the arguments of the CLI.

This is a perfet case of why it is so crucial to be able to manage all of those defaults by hand.

Reply:
> However, note that for SGD the learning rate by design is a required parameter and has no default. So in this case only giving the class path would not work

For the case of `SGD`, it would be okay to expect users to pass both `--optimizer=SGD` and `--optimizer.lr=0.001`

Reply:
They would be forced to always set that? As in there is no way for them to establish the default (in either the code or some sort of config file for defaults?). I would rather just handcoded everything so I can have defaults... The defaults established for the core pytorch classes will virtually always be the wrong ones for any particular application, so I think this is a more general issue not specific to optimizers and schedulers. 

For example how do you set defaults for the early stopping classes? It is very similar in thhat the defaults are always wrong and need to be established by the user

Reply:
> (in either the code or some sort of config file for defaults?).

I believe you can do something like

`$ python script.py --config=defaults.yaml --optimizer=SGD`

Where `defaults.yaml` contains all your custom defaults and any arguments to the right or it would override them.

Otherwise, in the code, you can override and call `parser.set_defaults({'optimizer.lr': 0.001})`

This is mentioned in the docs here: https://pytorch-lightning.readthedocs.io/en/latest/common/lightning_cli.html#configurable-callbacks

Reply:
That part of the docs for configurable callbacks doesn't have any examples using command line arguments, which I think is the source of a lot of my confusion. I am not sure what the commandline looks like to change one of those given the defaults (which you say can be in a defaults config file or in a subclasses LightningCLI. I am completely happy writing in yml itself if you prefer it that way.

So much what does overriding them look like on the commandline? Is it passing in a big dictionary which the user has to type out completely, or a more standard CLI. And do I need to provide all of the arguments for the underlying class or just the one I override?

I think examples here (Including in the docs) would help if you intend to support command line arguments.

Reply:
I will create a pull request with improvements to the documentation based on comments from @jlperla